further_link/util/bluetooth/server.py

```diff
@@ -0,0 +1,166 @@
+        await asyncio.sleep(2)
+        self.server.read_request_func = self._read_request
+        self.server.write_request_func = self._write_request
+        await self.server.add_gatt(self.config.tree)
+
+        await self.server.start()
+        logging.info("Bluetooth server is running")
+
+    async def stop(self) -> None:
+        if self.server:
+            logging.debug("Stopping bluetooth server")
+            await self.server.stop()
+
+    def _get_service_uuid(self, characteristic_uuid: str) -> Optional[str]:
+        if self.server is None:
+            raise Exception("Bluetooth server not started")
+        uuid = ""
+        try:
+            for service_uuid in self.server.services:
+                service = self.server.services[service_uuid]
+                if service.get_characteristic(characteristic_uuid):
+                    uuid = str(service.uuid)
+                    break
+        except Exception:
+            raise Exception(f"Unknown characteristic uuid: {characteristic_uuid}")
+
+        return uuid
+
+    def write_value(self, value: Union[str, bytearray], uuid: str) -> None:
+        """Write a characteristic value, notifying subscribers"""
+        if self.server is None:
+            raise Exception("Bluetooth server not started")
+
+        if isinstance(value, str):
+            value = to_byte_array(value)
+
+        logging.debug(f"Writing '{value}' to characteristic {uuid}")
+        characteristic = self.server.get_characteristic(uuid)
+        characteristic.value = value
+
+        # Notify subscribers
+        self.server.update_value(self._get_service_uuid(uuid), uuid)
+
+    def read_value(self, uuid: str) -> Optional[bytearray]:
+        if self.server is None:
+            raise Exception("Bluetooth server not started")
+
+        value = bytearray()
+        service_uuid = self._get_service_uuid(uuid)
+        service = self.server.services.get(service_uuid)
+        if service:
+            char = service.get_characteristic(uuid)
+            value = self._read_request(char)
+        return value
+
+    def _read_request(self, characteristic: BlessGATTCharacteristic, **kwargs):
+        """method required by bless to handle read requests"""
+
+        logging.debug(f"Read request for characteristc {characteristic}")
+        value = decode_value(characteristic.value)
+
+        # callback on read requests will update the value of the characteristic before returning
+        callback = (
+            self.config.tree.get(self._get_service_uuid(characteristic.uuid), {})
+            .get(characteristic.uuid, {})
+            .get("ReadHandler")
+        )
+        if callback and callable(callback):
+            logging.debug(f"Executing ReadHandler callback for {characteristic.uuid}")
+            try:
+                value = str(callback())
+                # TODO: handle long messages using ChunkedMessage
+                self.write_value(value, characteristic.uuid)
```

Comment:
let's try not writing here - we're not expecting people to be subscribed to notifs as well as reading so it shouldn't be necessary, just return the callback value